lib/open_food_network/proxy_order_syncer.rb

```diff
@@ -0,0 +1,79 @@
+module OpenFoodNetwork
+  class ProxyOrderSyncer
+    attr_reader :standing_order
+
+    delegate :order_cycles, :proxy_orders, :begins_at, :ends_at, to: :standing_order
+
+    def initialize(standing_orders)
+      case standing_orders
+      when StandingOrder
+        @standing_order = standing_orders
+      when ActiveRecord::Relation
+        @standing_orders = standing_orders.not_ended.not_canceled
+      else
+        raise "ProxyOrderSyncer must be initialized with an instance of StandingOrder or ActiveRecord::Relation"
+      end
+    end
+
+    def sync!
+      return sync_all! if @standing_orders
+      return initialise_proxy_orders! unless @standing_order.id
+      create_proxy_orders!
+      remove_obsolete_proxy_orders!
+    end
+
+    private
+
+    def sync_all!
+      @standing_orders.each do |standing_order|
+        @standing_order = standing_order
+        create_proxy_orders!
+        remove_obsolete_proxy_orders!
+      end
+    end
+
+    def initialise_proxy_orders!
+      uninitialised_order_cycle_ids.each do |order_cycle_id|
+        proxy_orders << ProxyOrder.new(standing_order: standing_order, order_cycle_id: order_cycle_id)
+      end
+    end
+
+    def create_proxy_orders!
+      return unless not_closed_in_range_order_cycles.any?
+      query = "INSERT INTO proxy_orders (standing_order_id, order_cycle_id, updated_at, created_at)"
+      query << " VALUES #{insert_values}"
+      query << " ON CONFLICT DO NOTHING"
```

Comment:
why do we need this?

Reply:
I wrote this to give a performance improvement. Depending on the data it can save on two or three queries per standing order. From memory it dropped the execution time for 150 StandingOrders x 10 ProxyOrders from 6~7 seconds down to about 0.5 seconds. I'll make a note to have another look at it and see if I think of another way of using ActiveRecord.

Reply:
More than that I was wondering about that "ON CONFLICT DO NOTHING"

Reply:
Each standing order has_many `ProxyOrders` which are just empty placeholders for actual `Order`s. When an `OrderCycle` is added to or removed from a Schedule, the list of ProxyOrders for each `StandingOrder` in the `Schedule` needs to be updated. List gives use a quick way of creating any new `ProxyOrders` that have not been initialised, while leaving existing ones without overwriting them.